src/llama-sampling.cpp

```diff
@@ -1059,6 +1059,86 @@ struct llama_sampler * llama_sampler_init_temp_ext(float temp, float delta, floa
     };
 }
 
+// xtc
+
+struct llama_sampler_xtc {
+    const float  probability;
+    const float  threshold;
+    const float  threshold_max;
+    const size_t min_keep;
+};
+
+static const char * llama_sampler_xtc_name(const struct llama_sampler * /*smpl*/) {
+    return "xtc";
+}
+
+static void llama_sample_xtc_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
+    const auto * ctx = (llama_sampler_xtc *) smpl->ctx;
+
+    if (ctx->probability <= 0.0f || ctx->threshold <= 0.0f || cur_p->size <= 1 || ctx->min_keep <= 2) {
+        return;
+    }
+
+    std::random_device rd;
+    float chance = (float)(rd()%100)/100;
+    if (chance > ctx->probability) return;
```

Comment:
I think I've resolved this now, although I may have misunderstood it. Are chances good now?

Reply:
`random_device` should only be used to initialize other RNGs, and we want the results to be repeatable when the seed is the same. You should keep a `std::mt19937` RNG in the sampler state in the same way that other samplers such as `llama_sampler_dist` do it. Take a `seed` parameter and reset the RNG usign this seed in the `reset` function. In this function, use the RNG in the state with `std::uniform_real_distribution` to obtain a number between 0 and 1.

Reply:
`reset` is only called when completely resetting the sampler to its original state, so the current implementation will only generate the `chance` once, which is probably not what you want. This is what I meant:

```diff
diff --git a/src/llama-sampling.cpp b/src/llama-sampling.cpp
index 4372b40c..2b368081 100644
--- a/src/llama-sampling.cpp
+++ b/src/llama-sampling.cpp
@@ -1069,9 +1069,8 @@ struct llama_sampler_xtc {

     const uint32_t seed;
     uint32_t       seed_cur;
-    float          chance;

-    std::mt19937 rng;
+    std::mt19937   rng;
 };

 static const char * llama_sampler_xtc_name(const struct llama_sampler * /*smpl*/) {
@@ -1079,7 +1078,7 @@ static const char * llama_sampler_xtc_name(const struct llama_sampler * /*smpl*/
 }

 static void llama_sample_xtc_apply(struct llama_sampler * smpl, llama_token_data_array * cur_p) {
-    const auto * ctx = (llama_sampler_xtc *) smpl->ctx;
+    auto * ctx = (llama_sampler_xtc *) smpl->ctx;

     if (ctx->probability <= 0.0f
         || ctx->threshold <= 0.0f
@@ -1090,8 +1089,10 @@ static void llama_sample_xtc_apply(struct llama_sampler * smpl, llama_token_data
         || ctx->min_keep <= 2) {
         return;
     }
-    // chance is calculated on init and on each reset
-    if (ctx->chance > ctx->probability) return;
+
+    std::uniform_real_distribution<float> distance(0.0f, 1.0f);
+    float chance = distance(ctx->rng);
+    if (chance > ctx->probability) return;

     // in case it's not sorted/recalculated yet
     llama_sampler_softmax_impl(cur_p);
@@ -1141,9 +1142,6 @@ static void llama_sampler_xtc_reset(struct llama_sampler * smpl) {
     auto * ctx = (llama_sampler_xtc *) smpl->ctx;
     ctx->seed_cur = get_rng_seed(ctx->seed);
     ctx->rng.seed(ctx->seed_cur);
-
-    std::uniform_real_distribution<> distance(0.0, 1.0);
-    ctx->chance = distance(ctx->rng);
 }

 static struct llama_sampler_i llama_sampler_xtc_i = {
@@ -1157,9 +1155,6 @@ static struct llama_sampler_i llama_sampler_xtc_i = {

 struct llama_sampler * llama_sampler_init_xtc(float p, float t, float t_max, size_t min_keep, uint32_t seed) {
     auto seed_cur = get_rng_seed(seed);
-    std::uniform_real_distribution<> distance(0.0, 1.0);
-    auto rng = std::mt19937(seed_cur);
-    float chance = distance(rng);
     return new llama_sampler {
         /* .iface = */ &llama_sampler_xtc_i,
         /* .ctx   = */ new llama_sampler_xtc {
@@ -1169,8 +1164,7 @@ struct llama_sampler * llama_sampler_init_xtc(float p, float t, float t_max, siz
             /* .min_keep      = */ min_keep,
             /* .seed          = */ seed,
             /* .seed_cur      = */ seed_cur,
-            /* .chance        = */ chance,
-            /* .rng           = */ rng,
+            /* .rng           = */ std::mt19937(seed_cur),
         },
     };
 }
```

Reply:
Oh, I assumed that `ctx` should stay `const` in `_apply`! Thanks, I definitely misunderstood that.